gh-120521: clarify except* documentation to allow tuples (#120523)
python/cpython: Doc/reference/compound_stmts.rst

until one is found that matches the exception.
An expression-less :keyword:`!except` clause, if present, must be last;
it matches any exception.

For an :keyword:`!except` clause with an expression, the
expression must evaluate to an exception type or a tuple of exception types.
The raised exception matches an :keyword:`!except` clause whose expression evaluates

> to the class or a :term:`non-virtual base class <abstract base class>` of the exception object,

Reviewer: This is a pre-existing problem and doesn't have to be solved in this PR, but I don't think linking to the definition of "abstract base class" here is great.

> or it evaluates to a tuple that contains such a class.

Reviewer: ```suggestion
or to a tuple that contains such a class.
```
I think this makes for a better sentence structure.

until one is found that matches the exception.
An expression-less :keyword:`!except` clause, if present, must be last;
it matches any exception.

For an :keyword:`!except` clause with an expression, the

> expression must evaluate to a type or a tuple containing types. Each type must
> be :exc:`BaseException` or a subclass of :exc:`BaseException`.

Reviewer 2: ```suggestion
expression must evaluate to an exception type or a tuple of exception types.
```

> be a subclass of :exc:`BaseException`, otherwise the runtime will raise a

Reviewer: ```suggestion
be a subclass of :exc:`BaseException`; otherwise, the runtime will raise a
```
Run-on sentence

> be a subclass of :exc:`BaseException`; otherwise, the runtime will raise a

Reviewer 2: It can be BaseException itself. Can this be worded to make that clear?

Author: clarified to "must be :exc:`BaseException` or a subclass of :exc:`BaseException`"

same for the BaseExceptionGroup mention in the except* section

:exc:`TypeError` when the handler is evaluated.

> The clause matches the exception if the resulting object is "compatible" with

Reviewer 2: ```suggestion
The clause matches an exception if the resulting object is "compatible" with
```

> The clause matches the exception if the resulting object is "compatible" with

Reviewer 2: Since you moved this sentence to a new paragraph, I don't think it's clear what "the resulting object" refers to.

Author: added a clarification to "resulting object from evaluating the clause"

> 
> The clause matches an exception if the resulting object from evaluating the
> clause is "compatible" with the exception. An object is compatible with an
> exception if the object is the class or a
> :term:`non-virtual base class <abstract base class>` of the exception
> object, or a tuple containing an item that is the class or a non-virtual base
> class of the exception object.

Reviewer 2: ```suggestion
The raised exception matches an :keyword:`!except` clause whose expression evaluates
to the class or a :term:`non-virtual base class <abstract base class>` of the exception object,
or it evaluates to a tuple that contains such a class.
```

   ...
   ExceptionGroup('', (BlockingIOError()))

> An :keyword:`!except*` clause must have a matching type or a tuple of matching

Reviewer: It might be worth expanding this further:

- The docs for `except` (the previous section) never say that each type must be a subtype of `BaseException`. That's probably worth adding.
- One important difference between `except` and `except*` is that the former allows "bare" except clauses (just `except:`), but `except*` doesn't. I think that's worth mentioning in a sentence by itself.
- We don't say what happens if the type is a subclass of BaseExceptionGroup.

So I'd suggest something like this:

```suggestion
Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a matching
expression; `except*:` is not valid syntax. When the handler is evaluated, the
expression must evaluate to a type or a tuple containing types. Each type must be a
subclass of :exc:`BaseException` and must not be a subclass of :exc:`BaseExceptionGroup`.
```

Author: Okay, I've added a mention that the runtime will raise a TypeError if the type is wrong, and updated the `except` section to have similar language.

Should the latter be split into a separate PR, or should I just update the title/summary of this PR?

   ...
   ExceptionGroup('', (BlockingIOError()))

Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a

> matching expression; ``except*:`` is not valid syntax. In addition to the

Reviewer 2: (I wouldn't say it's invalid syntax when it generates a runtime error).

Reviewer: It's a SyntaxError:
```
>>> try: x
... except *: pass
  File "<stdin>", line 2
    except *: pass
            ^
SyntaxError: expected one or more exception types
```

Reviewer 2: Ok, that one is but not this one:

```
>>> try: 1/0
... except* BaseExceptionGroup: pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: division by zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: catching ExceptionGroup with except* is not allowed. Use except instead.
```

> Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a
> matching expression; ``except*:`` is not valid syntax. In addition to the
> requirements for matching expressions in :keyword:`except` clauses, the
> types in the matching expression of an :keyword:`!except*` clause cannot
> be :exc:`BaseExceptionGroup` or a subclass of :exc:`BaseExceptionGroup`.

Reviewer 2: ```suggestion
An :keyword:`!except*` clause must have a matching expression; it cannot be ``except*:``.
Furthermore, this expression cannot contain exception group types, because that would
have ambiguous semantics.
```

> be :exc:`BaseExceptionGroup` or a subclass of :exc:`BaseExceptionGroup`.

Reviewer 2: (Above we talk about "exception group" when we refer to BaseException and its subclasses).

Author: Sorry, I'm confused about this comment - is it addressed in your suggestions already?

The other suggestions on this PR which I've merged in seem to have removed all references to BaseException and BaseExceptionGroup classnames in this doc. 

Feels like explicitly mentioning the class names would be more clear, wouldn't it?

Reviewer 2: Yes, this comment is addressed.

Since we talk about "exception group" in the previous paragraphs, assuming the reader knows what we mean there, I don't think specifying the type here would add clarity.

   ...
   ExceptionGroup('', (BlockingIOError()))

Unlike :keyword:`except` clauses, :keyword:`!except*` clauses must have a matching
expression; ``except*:`` is not valid syntax. When the handler is evaluated, the
expression must evaluate to a type or a tuple containing types. Each type must
be a subclass of :exc:`BaseException` and must not be a subclass of

> :exc:`BaseExceptionGroup`, otherwise the runtime will raise a :exc:`TypeError`

Reviewer: ```suggestion
:exc:`BaseExceptionGroup`; otherwise, the runtime will raise a :exc:`TypeError`
```
Same

> when the handler is evaluated.

Reviewer 2: Can we not repeat this twice? It is confusing because people assume it will somehow be different from the except case.

Author: Replaced the duplicate part with "In addition to the requirements for matching expressions in except clauses"

```
In addition to the
requirements for matching expressions in :keyword:`except` clauses, the
types in the matching expression of an :keyword:`!except*` clause cannot
be :exc:`BaseExceptionGroup` or a subclass of :exc:`BaseExceptionGroup`;
otherwise, the runtime will raise a :exc:`TypeError` when the handler is
evaluated.
```
